src/runner.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python3
+
+import io
+import os
+import shutil
+import subprocess
+import tarfile
+from pathlib import Path
+
+import requests
+
+from jinja2 import Environment, FileSystemLoader
+
+
+class Runner:
+    release_url = "https://api.github.com/repos/actions/runner/releases/latest"
+    runner_path = Path("/opt/github-runner")
+    env_file = runner_path / ".env"
+
+    def __init__(self):
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY", None)
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY", None)
+        self.proxies = {}
+        if http_proxy:
+            self.proxies["http"] = http_proxy
+        if https_proxy:
+            self.proxies["https"] = https_proxy
+        self.session = requests.Session()
+        if self.proxies:
+            self.session.proxies.update(self.proxies)
+        self.jinja = Environment(loader=FileSystemLoader("templates"))
+
+    def download(self):
```

Comment:
It's generally recommended to handle assets like this as resources, to ensure that all units are running the same version and to provide an explicit upgrade mechanism. That said, the runner by definition needs to have access to the GH API server and using that directly means we don't have to manage a copy of it.

Reply:
The runner also does automatic upgrades, so you can't guarantee the version since it will self update before it runs a job if it is behind.